### ISSUE TITLE
Working Group: Behavior and specification of URL redirects and renamings

### ISSUE THREAD

**Issue by bernhold** [betterscientificsoftware/bssw.io#704]
Charter: We have a couple of redirects, which have been implemented as one-offs.  We have the capability to manually override the automatically generated title-based slug use in a resource's URL.  For the long-term we need to figure out what really are our requirements and how we want to specify them.

Members: David, Pat, Rinku

**Comment by bernhold** [comment]
One think I think we *may*  need is the ability to have an article appear at multiple URLs.  Scenario:
1. We publish an article.  Maybe someone forwards the link to a friend or bookmarks it or something
2. We update the article in a way that argues for changing the title
3. If we change the name, the old links don't work

Possible solutions:
* Use the new metadata option to retain the original slug as the URL
  - In this case, the article is available *only* through the old URL
  - The new title and the URL are not connected.  Do we care?
* We modify the behavior so that the article appears at *both* the title's URL *and* the one from the metadata
  - Another approach would be to support the specification of multiple titles in the Markdown document.  The first would be the one that's visible on the published article.  All of the titles would be used to generate URL slugs at which the article would be available

**Comment by bernhold** [comment]
Consensus of initial discussions in 2021-03-11 EB meeting was ...
* Things controlling URL should be together in the article, not split out to other locations
* The initial URL under which something was published should be maintained
* Should be able to control whether "new" URL specifications add to or replace the title-generated URL
* Need to ensure that testing for title (URL) collisions includes *all* sources of URLs (i.e., metadata controls as well as actual titles)

We did not reach a consensus on whether an article should be available under multiple URLs if its title changes over time (i.e. initial URL and the URL based on the revised title).

**Comment by rinkug** [comment]
Lets separate functionality and implementation. 

I see two cases from **functionality perspective:**

**General** 
* The initial URL under which something was published should be maintained

**Article is updated:**
* If the title does not change, then URL remains the same and article is simply updated
* If title changes, then we need a rule such as "retain X popular past URLs which redirect to the latest title"..

**Article is not updated but we need custom URL:**
* If we need to specify custom URL (as we did for psip landing page), then we need redirection and that is independent of the title of the landing page

**For implementation** - now that has many solutions. Do we use metadata in each file, or a universal file? How do we detect collisions etc.

**Comment by bernhold** [comment]
@rinkug These two statements are not consistent:

> * The initial URL under which something was published should be maintained

> * If title changes, then we need a rule such as "retain X popular past URLs which redirect to the latest title"..

We would need to keep the **first** URL, the current URL, and perhaps some of the others.  How do we decide which ones?

A related question is: can we manage this manually (i.e. modify the metadata to keep the URLs we want)?  Or do we need the system to do it for us?  That significantly increases the complexity of the front-end (it requires memory spanning different versions of the content, and maybe some algorithm to determine which title changes are significant and which aren't).

An alternative proposal: We use *only* the URL under which it was originally published, in perpetuity.  No other URL will resolve to it, ever.

**Comment by bernhold** [comment]
@rinkug @pagrubel I would like to move this topic forward. I'd like to try to get together at our next operational meeting, if possible. (22 April?)

**Comment by bernhold** [comment]
Proposal: 

Add metadata field "Redirect-From:" and an argument of either a quoted title or a URL relative to https://bssw.io.  The title would be slugified to produce the URL in question.  Multiple arguments could be supplied.

We think this will suffice for https://bssw.io/psip style redirects too.

**Comment by bernhold** [comment]
#1294 has a specific redirect request from Elsa.  This has been implemented as what I'll call an alias (URL not rewritten).  We need to consider whether the Fellowship section of the site can be covered by our metadata-based solution.  That's the second part of the comment above.  I've asked PC to comment on this.

**Comment by markcmiller86** [comment]
Its not to relevant to this issue but I wanted to mention I just recently begun using [Jekyll's redirect functionality](https://superdevresources.com/redirects-jekyll-github-pages/) in another project and found it quite easy to use.

**Comment by bernhold** [comment]
Yeah, I've used it on occasion too.  [The proposal I made above](https://github.com/betterscientificsoftware/bssw.io/issues/704#issuecomment-1171496937) is similar in spirit and was inspired by the Jekyll capability.
